core/src/main/java/bisq/core/dao/state/DaoStateService.java

```diff
@@ -348,16 +360,16 @@ public Coin getGenesisTotalSupply() {
                 .flatMap(block -> block.getTxs().stream());
     }
 
-    public TreeMap<String, Tx> getTxMap() {
-        return new TreeMap<>(getTxStream().collect(Collectors.toMap(Tx::getId, tx -> tx)));
+    public Map<String, Tx> getTxMap() {
+        return daoState.getTxMap();
     }
 
     public Set<Tx> getTxs() {
-        return getTxStream().collect(Collectors.toSet());
+        return new HashSet<>(getTxMap().values());
```

Comment:
I switched it from getTxStream() to getTxMap.values() for slightly better performance, as more indirection is required to iterate through all the block tx lists (and many of the blocks are empty). They're not interchangeable elsewhere, though, since getTxMap.values() is unordered. It doesn't matter in this case since (before and after) it's just collecting into an unordered set.

Actually, the only place getTxs() is used is in one of the views to get the total number of transactions (using size()), so it doesn't need to collect into a new HashSet and the return type could just be a Collection<Tx>.

Reply:
These are the callers I found of `getTxStream()`. It seems like most/all of them can be satisfied by the cache, but you have been looking at it longer than me so I might be missing a use case.

```getTxs: size()
getInvalidTxs: size()
getIrregularTxs: size()
getTotalBurntFee: sum()
getTotalAmountOfInvalidatedBsq: sum()
getTotalAmountOfBurntBsq: sum()
getBurntFeeTxs: Set<Tx>
getTx: Optional<Tx>
getTxOutputStream: [boolean, Optional<TxOutput>, Set<TxOutput>

maybeExportToJson: <no idea if this needs to be ordered...>
```

My point at a high level was that if we are going to build a cache for something, we should just use it for everything. By keeping an interface and users for the old slow way, It just creates additional performance work in the future that could be avoided by doing something better now.

Adding javadocs as you change methods is also a great way to reduce the technical debt and help explain to future users the guarantees or any gotchas that you learned the hard way.

Reply:
I looked to all usages as well:
Here my comments:

`getInvalidTxs`
Used in UI for numTxs and in a `Set` for the chart. As it's a `Set` order is irrelevant.

`getIrregularTxs: size()`
Used in UI for numTxs -> order is irrelevant.

```
getTotalBurntFee: sum()
getTotalAmountOfInvalidatedBsq: sum()
getTotalAmountOfBurntBsq: sum()
```
Order is irrelevant for sum().

`getBurntFeeTxs: Set<Tx>`
Used in UI for numTxs and in a `Set` for the chart. As it's a `Set` order is irrelevant.

`getTx: Optional<Tx>`
TxId is unique so order is irrelevant.

`getTxOutputStream: [boolean, Optional<TxOutput>, Set<TxOutput>`
Used in `existsTxOutput` and `getTxOutput`:
Both use `txOutputKey`and as `txOutputKey` is unique tx order is irrelevant.

`maybeExportToJson`
It dumps txs and txOutputs to disk using the txId and txOutputKey as file name. As both are unique, order is irrelevant.

I guess the explorer will love to get some performance improvement from that change as well ;-). Writing to disk at each new block will be probably a major performance iussue there (another "low hanging fruit" -;) ).

Also looked into all usages of `block.getTxs()` and I do not see any write beside the addition of a tx in `onNewTxForLastBlock`. Maybe a public method inside `Block` encapsulating the write operation and exposing the getTxs() as unmodifiable List would make the code more safe. The read access cannot be replaced by the txMap, but as the write for both data structures happen at only one place I think it is safe (we do not have multiple threads here).

All the other data objects are immutable, only with `Block` it was hard as during parsing we need to know the state of past transaction in the same block. So keeping that local as we do with txs would have duplicate lots of code in DoaStateService where we need to look up if for instance BSQ is already spent by a previous tx.